### PR TITLE
doc: Initial NVMe-oF gateway docs

### PR DESCRIPTION
**Pull request by oritwas**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [x] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by zdover23** [comment]
@oritwas: May I make changes to this branch now? I see a few places where the formatting will cause the ordered list to render incorrectly, and experience has taught me that it's much easier for everyone if I just fix it instead of playing GitHub ping pong.

Thanks for this contribution!

**Comment by oritwas** [comment]
jenkins retest this please

**Comment by zdover23** [comment]
@oritwas: Should this be backported only to Squid? (Is this feature expected to land in Squid, I mean?)

**Comment by oritwas** [comment]
> @oritwas: May I make changes to this branch now? I see a few places where the formatting will cause the ordered list to render incorrectly, and experience has taught me that it's much easier for everyone if I just fix it instead of playing GitHub ping pong.
> 
> Thanks for this contribution!

**Comment by oritwas** [comment]
@zdover23 you can make any needed changes :)

**Comment by zdover23** [comment]
> @zdover23 you can make any needed changes :) 

@oritwas: Roger that. I'll make changes in a few hours, when I get back to work. Thanks for letting me know!

**Review by idryomov — COMMENTED** [review]
There is a [osd: remove extra '>' in spg_t Formatter](https://github.com/ceph/ceph/pull/55540/commits/6782d804714c3f92e2265149e4a58a7a0b685785) commit here that should go -- it's definitely not related, probably a rebase mishap.

**Comment by anthonyeleven** [comment]
> @oritwas: Should this be backported only to Squid? (Is this feature expected to land in Squid, I mean?)

The gateway's been under development for a few years, I suspect maybe it's time to roll it in with the core?

**Comment by zdover23** [comment]
> There is a [osd: remove extra '>' in spg_t Formatter](https://github.com/ceph/ceph/pull/55540/commits/6782d804714c3f92e2265149e4a58a7a0b685785) commit here that should go -- it's definitely not related, probably a rebase mishap.

@idryomov, I'll remove it if you're cool with the docs guy removing it. It's https://github.com/ceph/ceph/pull/55540/commits/6782d804714c3f92e2265149e4a58a7a0b685785, and I'm confident that I can interactively rebase against main and remove it.

**Comment by idryomov** [comment]
> @idryomov, I'll remove it if you're cool with the docs guy removing it. It's [6782d80](https://github.com/ceph/ceph/commit/6782d804714c3f92e2265149e4a58a7a0b685785), and I'm confident that I can interactively rebase against main and remove it.

Go for it!

**Comment by zdover23** [comment]
> > @idryomov, I'll remove it if you're cool with the docs guy removing it. It's [6782d80](https://github.com/ceph/ceph/commit/6782d804714c3f92e2265149e4a58a7a0b685785), and I'm confident that I can interactively rebase against main and remove it.
> 
> Go for it!

https://www.youtube.com/watch?v=UjvGAYuWSUA

**Comment by zdover23** [comment]
> > > @idryomov, I'll remove it if you're cool with the docs guy removing it. It's [6782d80](https://github.com/ceph/ceph/commit/6782d804714c3f92e2265149e4a58a7a0b685785), and I'm confident that I can interactively rebase against main and remove it.
> > 
> > 
> > Go for it!
> 
> https://www.youtube.com/watch?v=UjvGAYuWSUA

...I spoke too soon. It has already been removed. Dang.

**Comment by idryomov** [comment]
> It has already been removed.

It's still there -- the first of 13 commits in the PR.

**Comment by idryomov** [comment]
> of 13 commits in the PR

... which would be nice to squash before merging.

**Comment by zdover23** [comment]
> > of 13 commits in the PR
> 
> ... which would be nice to squash before merging.

Oh, I'm going to squash them. I don't merge docs changes unless they've been squashed into a single commit.

**Comment by anthonyeleven** [comment]
> Oh, I'm going to squash them. I don't merge docs changes unless they've been squashed into a single commit.

Snob.  😛

**Comment by zdover23** [comment]
> > It has already been removed.
> 
> It's still there -- the first of 13 commits in the PR.

Check this out:

**Comment by idryomov** [comment]
I think it's because that commit got merged to main as b121118e5a06f2acd5d9e5d5630af93c3847bee7.  So just the result of your rebase and that should do it.

**Comment by zdover23** [comment]
```
The following tests FAILED:
	 31 - run-rbd-unit-tests-1.sh (Failed)
	 32 - run-rbd-unit-tests-61.sh (Timeout)
```
jenkins retest this please

**Comment by zdover23** [comment]
jenkins retest this please

File: doc/rbd/nvmeof-initiator-linux.rst
Comment (idryomov) [inline doc/rbd/nvmeof-initiator-linux.rst:8]: This is upstream documentation, so it would be nice to avoid the impression that RHEL is required.

@oritwas What is the specific patchset/feature that requires RHEL 9.2 as opposed to e.g. any RHEL 9 or CentOS Stream 9?  What is the corresponding requirement for mainline kernels?

File: doc/rbd/nvmeof-initiator-linux.rst
Comment (idryomov) [inline doc/rbd/nvmeof-initiator-linux.rst:52]: My recollection is that bare `mkfs` creates ext2 filesystems by default.  I don't think we want to suggest that:
```suggestion
      mkfs.ext4 NVME_NODE_PATH
```

File: doc/rbd/nvmeof-initiator-linux.rst
Comment (idryomov) [inline doc/rbd/nvmeof-initiator-linux.rst:64]:
```suggestion
4. List the NVME-oF files:
```

File: doc/rbd/nvmeof-overview.rst
Comment (idryomov) [inline doc/rbd/nvmeof-overview.rst:8]:
```suggestion
RADOS Block Device (RBD) images as NVMe namespaces. The NVMe-oF protocol allows
```

File: doc/rbd/nvmeof-target-configure.rst
Comment (idryomov) [inline doc/rbd/nvmeof-target-configure.rst:6]: Not quite true -- this disregards the kernel client.  But this is just a nit, I don't insist on rewording this paragraph.

File: doc/rbd/nvmeof-target-configure.rst
Comment (idryomov) [inline doc/rbd/nvmeof-target-configure.rst:18]:
```suggestion
-  NVMe-oF gateways, which can either be colocated with OSD nodes or on dedicated nodes
```

File: doc/rbd/nvmeof-target-configure.rst
Comment (idryomov) [inline doc/rbd/nvmeof-target-configure.rst:25]:
```suggestion
The Ceph NVMe-oF gateway is both an NVMe-oF target and a Ceph client. Think of
```

File: doc/rbd/nvmeof-target-configure.rst
Comment (idryomov) [inline doc/rbd/nvmeof-target-configure.rst:38]: There is a Prerequisites section above that says the same thing with regards to the storage cluster but lists a different RHEL version.  Once versions are clarified, I think this section should be removed.

File: doc/rbd/nvmeof-target-configure.rst
Comment (idryomov) [inline doc/rbd/nvmeof-target-configure.rst:77]:
```suggestion
   podman run -it quay.io/ceph/nvmeof-cli-rhel9:latest --server-address GATEWAY_IP --server-port GATEWAY_PORT 5500 create_subsystem --subnqn SUSYSTEM_NQN --max-namespaces 256
```

File: doc/rbd/nvmeof-target-configure.rst
Comment (idryomov) [inline doc/rbd/nvmeof-target-configure.rst:99]: There is no `get` command.  Perhaps you meant `cat`?

File: doc/rbd/nvmeof-target-configure.rst
Comment (idryomov) [inline doc/rbd/nvmeof-target-configure.rst:109]: ditto

File: doc/rbd/nvmeof-target-configure.rst
Comment (oritwas) [reply doc/rbd/nvmeof-target-configure.rst:6]: It was copied for the iSCSi gateway docs

File: doc/rbd/nvmeof-initiator-esx.rst
Comment (zdover23) [reply doc/rbd/nvmeof-initiator-esx.rst:64]: Remedied.

File: doc/rbd/nvmeof-initiator-linux.rst
Comment (zdover23) [reply doc/rbd/nvmeof-initiator-linux.rst:52]: Remedied.

File: doc/rbd/nvmeof-overview.rst
Comment (zdover23) [reply doc/rbd/nvmeof-overview.rst:8]: Remedied.

File: doc/rbd/nvmeof-target-configure.rst
Comment (zdover23) [reply doc/rbd/nvmeof-target-configure.rst:18]: Remedied.

File: doc/rbd/nvmeof-target-configure.rst
Comment (zdover23) [reply doc/rbd/nvmeof-target-configure.rst:25]: Remedied.

File: doc/rbd/nvmeof-target-configure.rst
Comment (zdover23) [reply doc/rbd/nvmeof-target-configure.rst:77]: s/qauy/quay/ remedied. The CLI question is beyond my art, and remains open.

File: doc/rbd/nvmeof-target-configure.rst
Comment (zdover23) [reply doc/rbd/nvmeof-target-configure.rst:109]: Remedied.

File: doc/rbd/nvmeof-target-configure.rst
Comment (zdover23) [reply doc/rbd/nvmeof-target-configure.rst:115]: Remedied.

File: doc/rbd/nvmeof-target-configure.rst
Comment (idryomov) [inline doc/rbd/nvmeof-target-configure.rst:87]: - I don't think the upstream container has `-rhel9` suffix?
- missing `-`
```suggestion
          podman run -it quay.io/ceph/nvmeof-cli:latest --server-address GATEWAY_IP --server-port GATEWAY_PORT 5500 listener add --subsystem SUBSYSTEM_NQN --gateway-name GATEWAY_NAME --traddr GATEWAY_IP --trsvcid 4420
```

File: doc/rbd/nvmeof-target-configure.rst
Comment (idryomov) [inline doc/rbd/nvmeof-target-configure.rst:103]: ditto

File: doc/rbd/nvmeof-target-configure.rst
Comment (idryomov) [inline doc/rbd/nvmeof-target-configure.rst:109]: YYou? ;)
```suggestion
      podman run -it quay.io/ceph/nvmeof-cli:latest --server-address GATEWAY_IP --server-port GATEWAY_PORT 5500 subsystem list
```

File: doc/rbd/nvmeof-target-configure.rst
Comment (idryomov) [inline doc/rbd/nvmeof-target-configure.rst:115]: ditto

File: doc/rbd/nvmeof-target-configure.rst
Comment (idryomov) [inline doc/rbd/nvmeof-target-configure.rst:121]: ditto

**Review by idryomov — COMMENTED** [review]
@oritwas Zac is going to need input from you on RHEL and mainline kernel versions -- see outstanding comments.

File: doc/rbd/nvmeof-target-configure.rst
Comment (zdover23) [reply doc/rbd/nvmeof-target-configure.rst]: ``subsystem add`` has now replaced ``create_subsystem`` in the text.
``--subsystem`` has now replaced ``--subnqn`` in the text.

File: doc/rbd/nvmeof-target-configure.rst
Comment (zdover23) [reply doc/rbd/nvmeof-target-configure.rst]: ``cat``has replaced ``get`` in the text.

File: doc/rbd/nvmeof-target-configure.rst
Comment (zdover23) [reply doc/rbd/nvmeof-target-configure.rst:6]: > Not quite true -- this disregards the kernel client. But this is just a nit, I don't insist on rewording this paragraph.

@idryomov, if I add the kernel client to the list of agents that have traditionally enjoyed block-level access, will it make the sentence unambiguously true?

File: doc/rbd/nvmeof-target-configure.rst
Comment (zdover23) [reply doc/rbd/nvmeof-target-configure.rst]: This section has been removed from the text.

File: doc/rbd/nvmeof-initiator-linux.rst
Comment (oritwas) [reply doc/rbd/nvmeof-initiator-linux.rst]: Add the upstream kernel version and the other distros.
As for RHEL the NVMe/TCP initiator was GA in RHEL 9.0 but got broken in RHEL 9.1 :(
We highly recommend using the one from RHEL 9.2

File: doc/rbd/nvmeof-target-configure.rst
Comment (idryomov) [reply doc/rbd/nvmeof-target-configure.rst:6]: I'm thinking of something like:
> Traditionally, block-level access to a Ceph storage cluster has been limited to QEMU and ``librbd``, which is a key enabler for adoption within OpenStack, and the Linux kernel client.

And let's make the same edit in `doc/rbd/iscsi-targets.rst` where this was copied from.

Comment (idryomov) [reply doc/rbd/nvmeof-target-configure.rst:6]: > Is the following sentence correct?
> 
> > Traditionally, block-level access to a Ceph storage cluster has been limited to QEMU, `librbd`, and the Linux kernel client.

QEMU accesses Ceph through librbd, not by itself, so not quite correct.  The sentence that I proposed was meant to be parsed as:
> Traditionally, block-level access to a Ceph storage cluster has been limited to:
> 1. QEMU and librbd, which is a key enabler for adoption within OpenStack
> 2. Linux kernel client

I left the "which is a key enabler for adoption within OpenStack" bit in place because it's relevant to QEMU.

Comment (zdover23) [reply doc/rbd/nvmeof-target-configure.rst:6]: @idryomov Roger that. I understand now. I'm going to solve this with parentheses (which will likely irritate Anthony D'Atri, to whom I prospectively apologize while nonetheless persisting). Thanks for your clear explanation, Ilya.

File: doc/rbd/nvmeof-target-configure.rst
Comment (zdover23) [reply doc/rbd/nvmeof-target-configure.rst:103]: Accepted.

File: doc/rbd/nvmeof-target-configure.rst
Comment (zdover23) [reply doc/rbd/nvmeof-target-configure.rst:109]: Accepted.

File: doc/rbd/nvmeof-target-configure.rst
Comment (zdover23) [reply doc/rbd/nvmeof-target-configure.rst:121]: Accepted.

**Comment by zdover23** [comment]
> @oritwas Zac is going to need input from you on RHEL and mainline kernel versions -- see outstanding comments.

I've incorporated all of @idryomov's suggestions here.

**Review by zdover23 — APPROVED** [review]

File: doc/rbd/nvmeof-initiator-linux.rst
Comment (idryomov) [inline doc/rbd/nvmeof-initiator-linux.rst:8]: Where did kernel 4.19 came from?  AFAICT commit https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=3f2304f8c6d6ed97849057bd16fee99e434ca796 landed in kernel 5.0.

File: doc/rbd/nvmeof-initiator-linux.rst
Comment (idryomov) [inline doc/rbd/nvmeof-initiator-linux.rst:22]: Technically, this module is installed together with the kernel.  I'd suggest "add" or "load" here.

File: doc/rbd/nvmeof-initiator-linux.rst
Comment (idryomov) [inline doc/rbd/nvmeof-initiator-linux.rst:43]:
```suggestion
Verify that the initiator is set up correctly:
```

File: doc/rbd/nvmeof-overview.rst
Comment (idryomov) [inline doc/rbd/nvmeof-overview.rst:9]:
```suggestion
clients (initiators) to send NVMe commands to storage devices (targets) over a
```

File: doc/rbd/nvmeof-overview.rst
Comment (idryomov) [inline doc/rbd/nvmeof-overview.rst:13]:
```suggestion
Each NVMe-oF gateway consists of an `SPDK <https://spdk.io/>`_ NVMe-oF target
```

File: doc/rbd/nvmeof-target-configure.rst
Comment (idryomov) [inline doc/rbd/nvmeof-target-configure.rst:51]:
```suggestion
#. Deploy the NVMe-oF gateway daemons on a specific set of nodes:
```

File: doc/rbd/nvmeof-target-configure.rst
Comment (idryomov) [inline doc/rbd/nvmeof-target-configure.rst:71]:
```suggestion
   podman run -it quay.io/ceph/nvmeof-cli:latest --server-address GATEWAY_IP --server-port GATEWAY_PORT 5500 subsystem add --subsystem SUSYSTEM_NQN
```

File: doc/rbd/nvmeof-target-configure.rst
Comment (idryomov) [inline doc/rbd/nvmeof-target-configure.rst:77]: "namespace" definitely doesn't make sense here:
```suggestion
    a. On the install node, get the NVME-oF Gateway name:
```

File: doc/rbd/nvmeof-initiator-linux.rst
Comment (zdover23) [reply doc/rbd/nvmeof-initiator-linux.rst:43]: Good catch.

Repaired.

File: doc/rbd/nvmeof-overview.rst
Comment (zdover23) [reply doc/rbd/nvmeof-overview.rst:9]: Accepted.

File: doc/rbd/nvmeof-overview.rst
Comment (zdover23) [reply doc/rbd/nvmeof-overview.rst:13]: Fixed.

File: doc/rbd/nvmeof-target-configure.rst
Comment (zdover23) [reply doc/rbd/nvmeof-target-configure.rst:51]: Accepted.

File: doc/rbd/nvmeof-target-configure.rst
Comment (zdover23) [reply doc/rbd/nvmeof-target-configure.rst:71]: Accepted.

File: doc/rbd/nvmeof-target-configure.rst
Comment (zdover23) [reply doc/rbd/nvmeof-target-configure.rst]: Accepted.

File: doc/rbd/nvmeof-initiator-linux.rst
Comment (zdover23) [reply doc/rbd/nvmeof-initiator-linux.rst]: Accepted.

File: doc/rbd/nvmeof-initiator-linux.rst
Comment (zdover23) [reply doc/rbd/nvmeof-initiator-linux.rst]: Accepted. I made this "Kernel 5.10 or later".

File: doc/rbd/nvmeof-overview.rst
Comment (idryomov) [reply doc/rbd/nvmeof-overview.rst]: Not quite -- "NMVe" is still there.

File: doc/rbd/nvmeof-target-configure.rst
Comment (zdover23) [reply doc/rbd/nvmeof-target-configure.rst:65]: Finally remedied.

File: doc/rbd/nvmeof-overview.rst
Comment (zdover23) [reply doc/rbd/nvmeof-overview.rst]: I have removed it, at long last.

**Review by idryomov — APPROVED** [review]

**Comment by zdover23** [comment]
https://github.com/ceph/ceph/pull/55723 - Squid backport

**Comment by idryomov** [comment]
@zdover23 I think we also need a reef backport since it says
> Starting with the Ceph Reef release, block-level access has been expanded to offer standard NVMe/TCP support

**Comment by zdover23** [comment]
https://github.com/ceph/ceph/pull/55724 - Reef backport

File: doc/rbd/nvmeof-target-configure.rst
Comment (idryomov) [inline doc/rbd/nvmeof-target-configure.rst:73]: @zdover23 Looks like there is an issue with rendering on this line: it's treated as a part of the command and the list numbering is off as a consequence.  Could you please take a look?

https://docs.ceph.com/en/latest/rbd/nvmeof-target-configure/#configuration

Comment (zdover23) [reply doc/rbd/nvmeof-target-configure.rst:73]: @idryomov: https://github.com/ceph/ceph/pull/55729

**Pull request merged**
